// @babel/helper-fixtures/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = get;
exports.multiple = multiple;
exports.readFile = readFile;

var _cloneDeep = _interopRequireDefault(require("lodash/cloneDeep"));

var _clone = _interopRequireDefault(require("lodash/clone"));

var _extend = _interopRequireDefault(require("lodash/extend"));

var _semver = _interopRequireDefault(require("semver"));

var _path = _interopRequireDefault(require("path"));

var _fs = _interopRequireDefault(require("fs"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const nodeVersion = _semver.default.clean(process.version.slice(1));

function humanize(val, noext) {
  if (noext) val = _path.default.basename(val, _path.default.extname(val));
  return val.replace(/-/g, " ");
}

function tryResolve(module) {
  try {
    return require.resolve(module);
  } catch (e) {
    return null;
  }
}

function assertDirectory(loc) {
  if (!_fs.default.statSync(loc).isDirectory()) {
    throw new Error(`Expected ${loc} to be a directory.`);
  }
}

function shouldIgnore(name, blacklist) {
  if (blacklist && blacklist.indexOf(name) >= 0) {
    return true;
  }

  const ext = _path.default.extname(name);

  const base = _path.default.basename(name, ext);

  return name[0] === "." || ext === ".md" || base === "LICENSE" || base === "options";
}

const EXTENSIONS = [".js", ".mjs", ".ts", ".tsx"];

function findFile(filepath, allowJSON) {
  const matches = [];

  for (const ext of EXTENSIONS.concat(allowJSON ? ".json" : [])) {
    const name = filepath + ext;
    if (_fs.default.existsSync(name)) matches.push(name);
  }

  if (matches.length > 1) {
    throw new Error(`Found conflicting file matches: ${matches.join(", ")}`);
  }

  return matches[0];
}

function get(entryLoc) {
  const suites = [];
  let rootOpts = {};
  const rootOptsLoc = tryResolve(entryLoc + "/options");
  if (rootOptsLoc) rootOpts = require(rootOptsLoc);

  for (const suiteName of _fs.default.readdirSync(entryLoc)) {
    if (shouldIgnore(suiteName)) continue;
    const suite = {
      options: (0, _clone.default)(rootOpts),
      tests: [],
      title: humanize(suiteName),
      filename: entryLoc + "/" + suiteName
    };
    assertDirectory(suite.filename);
    suites.push(suite);
    const suiteOptsLoc = tryResolve(suite.filename + "/options");
    if (suiteOptsLoc) suite.options = require(suiteOptsLoc);

    for (const taskName of _fs.default.readdirSync(suite.filename)) {
      push(taskName, suite.filename + "/" + taskName);
    }

    function push(taskName, taskDir) {
      const taskDirStats = _fs.default.statSync(taskDir);

      let actualLoc = findFile(taskDir + "/input");
      let execLoc = findFile(taskDir + "/exec");

      if (taskDirStats.isDirectory() && !actualLoc && !execLoc) {
        if (_fs.default.readdirSync(taskDir).length > 0) {
          console.warn(`Skipped test folder with invalid layout: ${taskDir}`);
        }

        return;
      } else if (!actualLoc) {
        actualLoc = taskDir + "/input.js";
      } else if (!execLoc) {
        execLoc = taskDir + "/exec.js";
      }

      const expectLoc = findFile(taskDir + "/output", true) || taskDir + "/output.js";
      const stdoutLoc = taskDir + "/stdout.txt";
      const stderrLoc = taskDir + "/stderr.txt";

      const actualLocAlias = suiteName + "/" + taskName + "/" + _path.default.basename(actualLoc);

      const expectLocAlias = suiteName + "/" + taskName + "/" + _path.default.basename(actualLoc);

      let execLocAlias = suiteName + "/" + taskName + "/" + _path.default.basename(actualLoc);

      if (taskDirStats.isFile()) {
        const ext = _path.default.extname(taskDir);

        if (EXTENSIONS.indexOf(ext) === -1) return;
        execLoc = taskDir;
        execLocAlias = suiteName + "/" + taskName;
      }

      const taskOpts = (0, _cloneDeep.default)(suite.options);
      const taskOptsLoc = tryResolve(taskDir + "/options");
      if (taskOptsLoc) (0, _extend.default)(taskOpts, require(taskOptsLoc));
      const test = {
        optionsDir: taskOptsLoc ? _path.default.dirname(taskOptsLoc) : null,
        title: humanize(taskName, true),
        disabled: taskName[0] === ".",
        options: taskOpts,
        validateLogs: taskOpts.validateLogs,
        ignoreOutput: taskOpts.ignoreOutput,
        stdout: {
          loc: stdoutLoc,
          code: readFile(stdoutLoc)
        },
        stderr: {
          loc: stderrLoc,
          code: readFile(stderrLoc)
        },
        exec: {
          loc: execLoc,
          code: readFile(execLoc),
          filename: execLocAlias
        },
        actual: {
          loc: actualLoc,
          code: readFile(actualLoc),
          filename: actualLocAlias
        },
        expect: {
          loc: expectLoc,
          code: readFile(expectLoc),
          filename: expectLocAlias
        }
      };

      if (taskOpts.minNodeVersion) {
        const minimumVersion = _semver.default.clean(taskOpts.minNodeVersion);

        if (minimumVersion == null) {
          throw new Error(`'minNodeVersion' has invalid semver format: ${taskOpts.minNodeVersion}`);
        }

        if (_semver.default.lt(nodeVersion, minimumVersion)) {
          return;
        }

        delete taskOpts.minNodeVersion;
      }

      if (taskOpts.os) {
        let os = taskOpts.os;

        if (!Array.isArray(os) && typeof os !== "string") {
          throw new Error(`'os' should be either string or string array: ${taskOpts.os}`);
        }

        if (typeof os === "string") {
          os = [os];
        }

        if (!os.includes(process.platform)) {
          return;
        }

        delete taskOpts.os;
      }

      if (test.exec.code.indexOf("// Async.") >= 0) {
        return;
      }

      suite.tests.push(test);
      const sourceMappingsLoc = taskDir + "/source-mappings.json";

      if (_fs.default.existsSync(sourceMappingsLoc)) {
        test.sourceMappings = JSON.parse(readFile(sourceMappingsLoc));
      }

      const sourceMapLoc = taskDir + "/source-map.json";

      if (_fs.default.existsSync(sourceMapLoc)) {
        test.sourceMap = JSON.parse(readFile(sourceMapLoc));
      }

      const inputMapLoc = taskDir + "/input-source-map.json";

      if (_fs.default.existsSync(inputMapLoc)) {
        test.inputSourceMap = JSON.parse(readFile(inputMapLoc));
      }

      if (taskOpts.throws) {
        if (test.expect.code) {
          throw new Error("Test cannot throw and also return output code: " + expectLoc);
        }

        if (test.sourceMappings) {
          throw new Error("Test cannot throw and also return sourcemappings: " + sourceMappingsLoc);
        }

        if (test.sourceMap) {
          throw new Error("Test cannot throw and also return sourcemaps: " + sourceMapLoc);
        }
      }

      if (!test.validateLogs && (test.stdout.code || test.stderr.code)) {
        throw new Error("stdout.txt and stderr.txt are only allowed when the 'validateLogs' option is enabled: " + (test.stdout.code ? stdoutLoc : stderrLoc));
      }

      if (test.options.ignoreOutput) {
        if (test.expect.code) {
          throw new Error("Test cannot ignore its output and also validate it: " + expectLoc);
        }

        if (!test.validateLogs) {
          throw new Error("ignoreOutput can only be used when validateLogs is true: " + taskOptsLoc);
        }
      }

      delete test.options.validateLogs;
      delete test.options.ignoreOutput;
    }
  }

  return suites;
}

function multiple(entryLoc, ignore) {
  const categories = {};

  for (const name of _fs.default.readdirSync(entryLoc)) {
    if (shouldIgnore(name, ignore)) continue;

    const loc = _path.default.join(entryLoc, name);

    assertDirectory(loc);
    categories[name] = get(loc);
  }

  return categories;
}

function readFile(filename) {
  if (_fs.default.existsSync(filename)) {
    let file = _fs.default.readFileSync(filename, "utf8").trimRight();

    file = file.replace(/\r\n/g, "\n");
    return file;
  } else {
    return "";
  }
}
